Added optional chaining operator to safely handle undefined values

PradyumnChauhan opened the pull request:
## Overview


1. This PR fixes #20892
2. This PR does the following: Moving this.topicIdsToTopicName and related assignments outside the loop.
Adding a null check before pushing topicIdsToTopicName[topicId].
Ensuring this.topicNameToPrerequisiteTopicNames[currentTopicName] is always a valid array.
3. (For bug-fixing PRs only) The original bug occurred because: topicNameToPrerequisiteTopicNames[topicName] can be undefined, and we're doing .length

## Essential Checklist

* [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
* [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
* [x] I have written tests for my code.
* [x] The **PR title** starts with "Fix #20892: " added a null check before pushing to prerequisiteTopicNames.
* [x] I have assigned the correct reviewers to this PR @HardikGoyal2003  

## Proof that changes are correct

https://github.com/user-attachments/assets/7cea9216-252e-4867-92ae-8128ce724c4c

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

PradyumnChauhan committed:
Update requirements_dev.txt after running install script

PradyumnChauhan committed:
Update requirements_dev.txt after running install script second time

PradyumnChauhan committed:
Improve topic dependency setup with null checks and structure cleanup

Vir-8 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -584,26 +584,28 @@ export class ClassroomAdminPageComponent implements OnInit {
         this.topicNameToPrerequisiteTopicNames = {};
+        this.tempClassroomData.setTopicIdToTopicName(topicIdsToTopicName);
+        this.topicIdsToTopicName = topicIdsToTopicName;
+        this.topicNames = Object.values(this.topicIdsToTopicName);
```
Not sure if I understand your approach, what difference does moving these calls to be before the loop make?

Vir-8 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -584,26 +584,28 @@ export class ClassroomAdminPageComponent implements OnInit {
-
-          this.topicNameToPrerequisiteTopicNames[currentTopicName] =
+          // Fallback to empty array if undefined.
+          this.topicNameToPrerequisiteTopicNames[currentTopicName || ''] =
```
If the currentTopicName is undefined, shouldn't nothing be changed in `topicNameToPrerequisiteTopicNames`?

Vir-8 reviewed (changes requested):
Hi @PradyumnChauhan, thanks for the PR!

I don't think I understand the fix, can you clarify why topic IDs are undefined in the first place?

PradyumnChauhan committed:
revised pr : handled undefined values safely

PradyumnChauhan commented:
hey @Vir-8 @HardikGoyal2003 ,
The `getPrerequisiteLength` function was throwing "TypeError: Cannot read properties of undefined (reading 'length')" when trying to access length of undefined prerequisite topics in `classroom-admin-page.component.ts`

fix:
Added optional chaining operator to safely handle undefined values:
and also removed the previous wrong commit..
```typescript

getPrerequisiteLength(topicName: string): number {
  return this.topicNameToPrerequisiteTopicNames[topicName]?.length || 0;
}
```

https://github.com/user-attachments/assets/82142be6-9862-4dfc-9fc5-f462dde553fa

PradyumnChauhan commented:
Hey @imchristie , can you please review this Pr!!

imchristie commented:
@PradyumnChauhan Not sure why you are asking for my review since this is not acceptance test related.

PradyumnChauhan commented:
Hi @imchristie  ,
Apologies for tagging you here — I now realize this PR isn’t related to acceptance tests.

Could you kindly let me know if there’s someone else more appropriate who could review this PR? I’d appreciate any guidance.

Vir-8 commented on core/templates/pages/classroom-admin-page/classroom-admin-page.component.ts:
```diff
@@ -861,7 +861,7 @@ export class ClassroomAdminPageComponent implements OnInit {
 
   getPrerequisiteLength(topicName: string): number {
-    return this.topicNameToPrerequisiteTopicNames[topicName].length;
+    return this.topicNameToPrerequisiteTopicNames[topicName]?.length;
```
This does not fix it, why is the prerequisite topic undefined in the first place?

If there is no prerequisite topic, we shouldn't even be trying to access its length. And I believe this function will still return undefined in that case, which will cause unexpected problems wherever the function is being called.

Vir-8 reviewed (commented):
Hi @PradyumnChauhan, thanks again for the PR!

Just a quick note: please avoid resolving comments on your own -- this makes it harder for reviewers to keep track. Also if you're unsure about how to get a reviewer assigned, please refer to our [PR rules](https://github.com/oppia/oppia/wiki/Rules-for-making-PRs).

I'm closing this PR for now, as the direction of the fix has changed significantly and it's unclear what the final approach is meant to be.

Feel free to submit a new PR in the future when you have a more clearer and consistent solution. Thanks!
